src/Common/assert_cast.h

```diff
@@ -25,7 +25,7 @@ namespace DB
 template <typename To, typename From>
 inline To assert_cast(From && from)
 {
-#ifndef NDEBUG
+#ifdef ABORT_ON_LOGICAL_ERROR
```

Comment:
it is basically equivalent to `debug or sanitizers`:

https://github.com/ClickHouse/ClickHouse/blob/97a5de96533048bb3503c1e3e8ad52416f2c8072/base/base/defines.h#L90-L94

Reply:
we could rename it to something like `DEBUG_OR_SANITIZER`
I see we have many places with `ABORT_ON_LOGICAL_ERROR` where in fact we are not aborting on logical error.

Reply:
> we could rename it to something like DEBUG_OR_SANITIZER
I see we have many places with ABORT_ON_LOGICAL_ERROR where in fact we are not aborting on logical error.

I had the same thoughts, but I didn't wanted to push this to @nikitat. Even though if you are up for it, feel free to rename it. Otherwise I can do it in a separate PR.